File: app/src/main/java/com/hazelmobile/filetransfer/adapter/ApplicationListAdapter.java
package com.hazelmobile.filetransfer.adapter;

import android.content.Context;
import android.content.SharedPreferences;
import android.content.pm.ApplicationInfo;
import android.content.pm.PackageInfo;
import android.content.pm.PackageManager;
import android.net.Uri;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.annotation.NonNull;

import com.hazelmobile.filetransfer.GlideApp;
import com.hazelmobile.filetransfer.R;
import com.hazelmobile.filetransfer.files.FileUtils;
import com.hazelmobile.filetransfer.pictures.EditableListAdapter;
import com.hazelmobile.filetransfer.pictures.Shareable;

import java.io.File;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class ApplicationListAdapter
        extends EditableListAdapter<ApplicationListAdapter.PackageHolder, EditableListAdapter.EditableViewHolder> {
    private SharedPreferences mPreferences;
    private PackageManager mManager;

    public ApplicationListAdapter(Context context, SharedPreferences preferences) {
        super(context);
        mPreferences = preferences;
        mManager = context.getPackageManager();
    }

    @Override
    public List<PackageHolder> onLoad() {
        List<PackageHolder> list = new ArrayList<>();
        boolean showSystemApps = mPreferences.getBoolean("show_system_apps", false);

        for (PackageInfo packageInfo : getContext().getPackageManager().getInstalledPackages(PackageManager.GET_META_DATA)) {
            ApplicationInfo appInfo = packageInfo.applicationInfo;

            if ((appInfo.flags & ApplicationInfo.FLAG_SYSTEM) != 1 || showSystemApps) {
                PackageHolder packageHolder = new PackageHolder(appInfo.loadLabel(mManager).toString(),
                        appInfo,
                        packageInfo.versionName,
                        packageInfo.packageName,
                        new File(appInfo.sourceDir),
                        FileUtils.sizeExpression(new File(appInfo.sourceDir).length(), false));

                if (filterItem(packageHolder))
                    list.add(packageHolder);
            }
        }

        Collections.sort(list, getDefaultComparator());

        return list;
    }

    @NonNull
    @Override
    public EditableListAdapter.EditableViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        return new EditableListAdapter.EditableViewHolder(getInflater().inflate(R.layout.apps_item_layout, parent, false));
    }

    @Override
    public void onBindViewHolder(@NonNull final EditableListAdapter.EditableViewHolder holder, final int position) {
        try {
            final View parentView = holder.getView();
            final PackageHolder object = getItem(position);
            ImageView image = parentView.findViewById(R.id.appIcon);
            TextView text1 = parentView.findViewById(R.id.appName);
            TextView text2 = parentView.findViewById(R.id.appSize);

            text1.setText(object.friendlyName);
            text2.setText(object.appSize);

            parentView.setSelected(object.isSelectableSelected());

            GlideApp.with(getContext())
                    .load(object.appInfo)
                    .override(160)
                    .centerCrop()
                    .into(image);
        } catch (Exception e) {
        }
    }

    public static class PackageHolder extends Shareable {
        public static final String FORMAT = ".apk";
        public static final String MIME_TYPE = FileUtils.getFileContentType(FORMAT);

        public ApplicationInfo appInfo;
        public String version;
        public String packageName;
        public String appSize;

        public PackageHolder(String friendlyName, ApplicationInfo appInfo, String version,
                             String packageName, File executableFile, String appSize) {
            super(appInfo.packageName.hashCode(),
                    friendlyName,
                    friendlyName + "_" + version + ".apk",
                    MIME_TYPE,
                    executableFile.lastModified(),
                    executableFile.length(),
                    Uri.fromFile(executableFile));

            this.appInfo = appInfo;
            this.version = version;
            this.packageName = packageName;
            this.appSize = appSize;
        }
    }
}

File: app/src/main/java/com/hazelmobile/filetransfer/ui/adapter/ApplicationListAdapter.kt
package com.hazelmobile.filetransfer.ui.adapter

import android.content.Context
import android.content.SharedPreferences
import android.content.pm.ApplicationInfo
import android.content.pm.PackageInfo
import android.content.pm.PackageManager
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import androidx.annotation.NonNull
import androidx.recyclerview.widget.RecyclerView
import com.genonbeta.android.framework.util.FileUtils
import com.hazelmobile.filetransfer.GlideApp
import com.hazelmobile.filetransfer.R
import com.hazelmobile.filetransfer.model.PackageHolder
import com.hazelmobile.filetransfer.pictures.EditableListAdapter
import com.hazelmobile.filetransfer.widget.getLogInfo
import kotlinx.android.synthetic.main.apps_item_layout.view.*
import java.io.File
import java.security.cert.Extension
import java.util.*

open class ApplicationListAdapter(private val mContext: Context, val preferences: SharedPreferences) :
    EditableListAdapter<PackageHolder, EditableListAdapter.EditableViewHolder>(mContext) {

    override fun onLoad(): MutableList<PackageHolder> {
        val appList = ArrayList<PackageHolder>()

        for (packageInfo: PackageInfo in mContext.packageManager.getInstalledPackages(PackageManager.GET_META_DATA)) {
            val appInfo: ApplicationInfo = packageInfo.applicationInfo
            if (appInfo.flags and ApplicationInfo.FLAG_SYSTEM != 1) {

                val packageHolder = PackageHolder(
                    appInfo.loadLabel(mContext.packageManager).toString(),
                    appInfo,
                    packageInfo.versionName,
                    packageInfo.packageName,
                    File(appInfo.sourceDir),
                    FileUtils.sizeExpression(File(appInfo.sourceDir).length(), false)
                )

                if (filterItem(packageHolder))
                    appList.add(packageHolder)
            }
        }

        Collections.sort(appList, defaultComparator)

        return appList
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): EditableViewHolder {
        return EditableViewHolder(inflater.inflate(R.layout.list_application, parent, false))
    }

    override fun onBindViewHolder(@NonNull holder: EditableViewHolder, position: Int) {
        try {
            val parentView = holder.view
            val packageHolder: PackageHolder = getItem(position)
            parentView.appName.text = packageHolder.appfriendlyName
            parentView.appSize.text = packageHolder.appSize
            GlideApp.with(context)
                .load(packageHolder.appInfo)
                .override(80)
                .centerCrop()
                .into(parentView.appIcon)
        } catch (exp: Exception) {
            getLogInfo(exp.message!!)
        }
    }

}